### PR TITLE
Allow unused params in CSIS inference

### PR DESCRIPTION
**Pull request by fritzo**
Addresses a comment on the forum
https://forum.pyro.ai/t/how-to-turn-on-allow-unused-true-through-pyro/1725

### CODE REVIEW

File: pyro/infer/csis.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def loss_and_grads(self, grads, batch, *args, **kwargs):
             if grads:
                 guide_params = set(site["value"].unconstrained()
                                    for site in particle_param_capture.trace.nodes.values())
-                guide_grads = torch.autograd.grad(particle_loss, guide_params)
+                guide_grads = torch.autograd.grad(particle_loss, guide_params, allow_unused=True)
```
Comment (neerajprad) [inline pyro/infer/csis.py]: Looks fine. Just to confirm the issue - this shouldn't happen unless `guide_params` contains parameters that aren't used in the computation of `particle_loss`. Is that possible or common?

**Comment by fritzo** [comment]
@DexiongYung does this fix your issue?

**Comment by DexiongYung** [comment]
I believe so

File: pyro/infer/csis.py (hunk 1 above)
Comment (fritzo) [reply pyro/infer/csis.py]: It is certainly possible e.g. in models with dynamic control flow with a dead branch, e.g.
```py
def model(data):
    a = pyro.param("a", torch.tensor(+1.))
    b = pyro.param("b", torch.tensor(-1.))
    x = pyro.sample("x", dist.Bernoulli(0.5))
    pyro.sample("y", dist.Normal(a if x else b, 1.), obs=data)
```
I don't recall having seen this error in any recent model, so I'm not sure how common it is.

Comment (neerajprad) [reply pyro/infer/csis.py]: Thanks for the snippet, makes sense.

**Pull request merged**